File: palettier-core/src/test/java/be/asmolabs/palettier/core/service/DryingTimeServiceTest.java
package be.asmolabs.palettier.core.service;

import static org.assertj.core.api.Assertions.assertThat;

import be.asmolabs.palettier.core.domain.DryingClass;
import be.asmolabs.palettier.core.domain.LayerThickness;
import be.asmolabs.palettier.core.domain.Medium;
import be.asmolabs.palettier.core.domain.Ventilation;
import be.asmolabs.palettier.core.service.DryingModels.DryingContext;
import be.asmolabs.palettier.core.service.DryingModels.DryingEstimate;
import be.asmolabs.palettier.core.service.DryingModels.Workshop;
import java.time.Duration;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

class DryingTimeServiceTest {

    private final DryingTimeService service = new DryingTimeService(new DryingProperties());

    private static DryingContext context(Medium medium, double ratio, LayerThickness thickness, Workshop workshop) {
        return new DryingContext(DryingClass.MEDIUM, medium, ratio, thickness, workshop);
    }

    @Test
    @DisplayName("les jalons de sechage sont ordonnes du temps ouvert a la polymerisation")
    void milestonesAreOrdered() {
        DryingEstimate estimate = service.estimate(
                context(Medium.ODORLESS_THINNER, 0.8, LayerThickness.GLAZE, Workshop.standard()));

        assertThat(estimate.openTime()).isLessThan(estimate.touchDry());
        assertThat(estimate.touchDry()).isLessThan(estimate.recoat());
        assertThat(estimate.recoat()).isLessThan(estimate.throughDry());
        assertThat(estimate.throughDry()).isLessThan(estimate.fullCure());
    }

    @Test
    @DisplayName("le diluant accelere le sechage, l'huile de lin le ralentit")
    void mediumChangesDryingSpeed() {
        Duration pure = touchDry(Medium.NONE, 0.0);
        Duration thinned = touchDry(Medium.ODORLESS_THINNER, 0.8);
        Duration oiled = touchDry(Medium.LINSEED_OIL, 0.4);

        assertThat(thinned).isLessThan(pure);
        assertThat(oiled).isGreaterThan(pure);
    }

    @Test
    @DisplayName("une couche epaisse seche nettement plus lentement qu'un glacis")
    void thicknessDominatesDryingTime() {
        Duration glaze = touchDry(LayerThickness.GLAZE);
        Duration impasto = touchDry(LayerThickness.IMPASTO);

        assertThat(impasto).isGreaterThan(glaze.multipliedBy(5));
    }

    @Test
    @DisplayName("dix degres de moins doublent le temps de sechage")
    void temperatureHalvesOrDoublesDryingTime() {
        Duration warm = touchDry(new Workshop(20, 50, Ventilation.NORMAL));
        Duration cold = touchDry(new Workshop(10, 50, Ventilation.NORMAL));

        assertThat(cold.toMinutes()).isCloseTo(warm.toMinutes() * 2, org.assertj.core.data.Offset.offset(2L));
    }

    @Test
    @DisplayName("une dilution excessive declenche un avertissement")
    void excessiveThinningIsReported() {
        DryingEstimate estimate = service.estimate(
                context(Medium.ODORLESS_THINNER, 0.98, LayerThickness.GLAZE, Workshop.standard()));

        assertThat(estimate.advice()).anySatisfy(a -> assertThat(a).contains("limite utile"));
    }

    @Test
    @DisplayName("un atelier froid est signale au peintre")
    void coldWorkshopIsReported() {
        DryingEstimate estimate = service.estimate(
                context(Medium.NONE, 0.0, LayerThickness.NORMAL, new Workshop(12, 50, Ventilation.NORMAL)));

        assertThat(estimate.advice()).anySatisfy(a -> assertThat(a).contains("15 degres"));
    }

    private Duration touchDry(Medium medium, double ratio) {
        return service.estimate(context(medium, ratio, LayerThickness.NORMAL, Workshop.standard())).touchDry();
    }

    private Duration touchDry(LayerThickness thickness) {
        return service.estimate(context(Medium.NONE, 0.0, thickness, Workshop.standard())).touchDry();
    }

    private Duration touchDry(Workshop workshop) {
        return service.estimate(context(Medium.NONE, 0.0, LayerThickness.NORMAL, workshop)).touchDry();
    }
}

File: palettier-kmp/core-domain/src/commonTest/kotlin/be/asmolabs/palettier/domain/drying/DryingTimeServiceTest.kt
package be.asmolabs.palettier.domain.drying

import be.asmolabs.palettier.domain.paint.DryingClass
import be.asmolabs.palettier.domain.paint.LayerThickness
import be.asmolabs.palettier.domain.paint.Medium
import be.asmolabs.palettier.domain.paint.Ventilation
import kotlin.math.abs
import kotlin.test.Test
import kotlin.test.assertTrue
import kotlin.time.Duration

/** Traduction fidele de DryingTimeServiceTest cote Java, valeurs comprises. */
class DryingTimeServiceTest {

    private val service = DryingTimeService()

    private fun context(
        medium: Medium = Medium.NONE,
        ratio: Double = 0.0,
        thickness: LayerThickness = LayerThickness.NORMAL,
        workshop: Workshop = Workshop.standard(),
    ) = DryingContext(DryingClass.MEDIUM, medium, ratio, thickness, workshop)

    private fun touchDry(
        medium: Medium = Medium.NONE,
        ratio: Double = 0.0,
        thickness: LayerThickness = LayerThickness.NORMAL,
        workshop: Workshop = Workshop.standard(),
    ): Duration = service.estimate(context(medium, ratio, thickness, workshop)).touchDry

    @Test
    fun `les jalons de sechage sont ordonnes du temps ouvert a la polymerisation`() {
        val estimate = service.estimate(
            context(Medium.ODORLESS_THINNER, 0.8, LayerThickness.GLAZE)
        )

        assertTrue(estimate.openTime < estimate.touchDry)
        assertTrue(estimate.touchDry < estimate.recoat)
        assertTrue(estimate.recoat < estimate.throughDry)
        assertTrue(estimate.throughDry < estimate.fullCure)
    }

    @Test
    fun `le diluant accelere le sechage, l'huile de lin le ralentit`() {
        val pure = touchDry(Medium.NONE, 0.0)
        val thinned = touchDry(Medium.ODORLESS_THINNER, 0.8)
        val oiled = touchDry(Medium.LINSEED_OIL, 0.4)

        assertTrue(thinned < pure)
        assertTrue(oiled > pure)
    }

    @Test
    fun `une couche epaisse seche nettement plus lentement qu'un glacis`() {
        val glaze = touchDry(thickness = LayerThickness.GLAZE)
        val impasto = touchDry(thickness = LayerThickness.IMPASTO)

        assertTrue(impasto > glaze * 5)
    }

    @Test
    fun `dix degres de moins doublent le temps de sechage`() {
        val warm = touchDry(workshop = Workshop(20.0, 50.0, Ventilation.NORMAL))
        val cold = touchDry(workshop = Workshop(10.0, 50.0, Ventilation.NORMAL))

        assertTrue(abs(cold.inWholeMinutes - warm.inWholeMinutes * 2) <= 2)
    }

    @Test
    fun `une dilution excessive declenche un avertissement`() {
        val estimate = service.estimate(context(Medium.ODORLESS_THINNER, 0.98, LayerThickness.GLAZE))
        assertTrue(estimate.advice.any { "limite utile" in it }, estimate.advice.toString())
    }

    @Test
    fun `un atelier froid est signale au peintre`() {
        val estimate = service.estimate(
            context(workshop = Workshop(12.0, 50.0, Ventilation.NORMAL))
        )
        assertTrue(estimate.advice.any { "15 degres" in it }, estimate.advice.toString())
    }

    @Test
    fun `la regle du gras sur maigre lit la proportion de medium`() {
        // Un diluant maigrit, une huile engraisse, la peinture pure est la reference.
        assertTrue(Medium.ODORLESS_THINNER.fatness(0.8) < Medium.NONE.fatness(0.0))
        assertTrue(Medium.LINSEED_OIL.fatness(0.4) > Medium.NONE.fatness(0.0))
    }
}
